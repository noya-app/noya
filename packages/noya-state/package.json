{
  "name": "noya-state",
  "version": "1.0.0",
  "main": "src/index.ts",
  "license": "UNLICENSED",
  "dependencies": {
    "immer": "^9.0.5",
    "tree-visit": "^0.4.0"
  }
}
