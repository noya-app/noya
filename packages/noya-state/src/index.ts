import { setAutoFreeze } from 'immer';

setAutoFreeze(false);

export * as Overrides from './overrides';
export * as Layers from './layers';
export * as GroupLayouts from './groupLayouts';
export * as ExportOptions from './exportOptions';
export * as Selectors from './selectors/selectors';
export * as Models from './models';
export * as TextStyleSelectors from './selectors/textStyleSelectors';

export type { SelectedPoint } from './reducers/pointReducer';
export * from './reducers/workspaceReducer';
export * from './reducers/applicationReducer';
export * from './reducers/historyReducer';
export * from './reducers/interactionReducer';
export * from './types';

export type { SelectionType } from './utils/selection';

export * as Primitives from './primitives';
export * from './primitives';
export * from './selection';
export * from './checkeredBackground';
export * from './snapping';
