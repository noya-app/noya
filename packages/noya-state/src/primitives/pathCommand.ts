import { TupleOf } from '@noya-app/noya-utils';

export enum PathCommandVerb {
  move = 0,
  line = 1,
  quad = 2,
  conic = 3,
  cubic = 4,
  close = 5,
}

const commandParameterCount = {
  [PathCommandVerb.move]: 2,
  [PathCommandVerb.line]: 2,
  [PathCommandVerb.quad]: 4,
  [PathCommandVerb.conic]: 5,
  [PathCommandVerb.cubic]: 6,
  [PathCommandVerb.close]: 0,
} as const;

export type PathCommand = {
  [P in PathCommandVerb]: [
    P,
    ...TupleOf<number, (typeof commandParameterCount)[P]>,
  ];
}[PathCommandVerb];

export function parsePathCmds(input: Float32Array) {
  const result: PathCommand[] = [];

  let i = 0;
  while (i < input.length) {
    const cmd = input[i] as PathCommandVerb;
    i++;
    const count = commandParameterCount[cmd];
    const params = input.slice(i, i + count);
    i += count;
    result.push([cmd, ...params] as PathCommand);
  }

  return result;
}
