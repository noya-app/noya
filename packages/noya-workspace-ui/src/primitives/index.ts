export * as Primitives from './primitives';
