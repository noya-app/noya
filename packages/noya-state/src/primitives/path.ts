import { Sketch } from '@noya-app/noya-file-format';
import { Point, Rect, distance } from '@noya-app/noya-geometry';
import { PointString, SketchModel } from '@noya-app/noya-sketch-model';
import { clamp, rotate, windowsOf, zip } from '@noya-app/noya-utils';
import { CanvasKit, Path, PathOp } from 'canvaskit';
import {
  CommandWithoutQuadratics,
  makePathsFromCommands,
} from 'noya-import-svg';
import {
  lineCapStyle,
  lineJoinStyle,
  parsePoint,
  stringifyPoint,
} from 'noya-state';
import { PathCommand, PathCommandVerb, parsePathCmds } from './pathCommand';

/**
 * The radius of an edge should be less than half the length of that edge
 */
function getEdgeRadius(a: Point, b: Point, cornerRadius: number): number {
  const dist = distance(a, b);

  return dist <= cornerRadius * 2 ? dist / 2 : cornerRadius;
}

/**
 * The radius of a corner should be symmetric on both sides
 */
function getCornerRadius(
  before: Point,
  corner: Point,
  after: Point,
  cornerRadius: number,
): number {
  const beforeRadius = getEdgeRadius(corner, before, cornerRadius);
  const afterRadius = getEdgeRadius(corner, after, cornerRadius);
  return Math.min(beforeRadius, afterRadius);
}

// Implementation loosely follows:
// https://github.com/google/skia/blob/13d6c4f55db0d1429f889704ed9b2e48968ac7a3/src/effects/SkCornerPathEffect.cpp#L24
function getStep(
  a: Point,
  b: Point,
  radius: number,
): { step: Point; drawSegment: boolean } {
  let dist = distance(a, b);

  const step: Point = { x: b.x - a.x, y: b.y - a.y };

  // Since we clamp radius beforehand, the dist will never be < radius * 2, only equal
  if (dist <= radius * 2) {
    step.x /= 2;
    step.y /= 2;

    return { step, drawSegment: false };
  }

  step.x *= radius / dist;
  step.y *= radius / dist;

  return { step, drawSegment: true };
}

export type DecodedCurvePoint = Omit<
  Sketch.CurvePoint,
  'curveFrom' | 'curveTo' | 'point'
> & {
  curveFrom: Point;
  curveTo: Point;
  point: Point;
};

export function scalePoint(point: Point, { x, y, width, height }: Rect): Point {
  return { x: x + point.x * width, y: y + point.y * height };
}

export function unscalePoint(
  point: Point,
  { x, y, width, height }: Rect,
): Point {
  return { x: (point.x - x) / width, y: (point.y - y) / height };
}

export function decodeCurvePoint(
  curvePoint: Sketch.CurvePoint,
  frame: Rect,
): DecodedCurvePoint {
  return {
    ...curvePoint,
    hasCurveFrom: curvePoint.hasCurveTo,
    hasCurveTo: curvePoint.hasCurveFrom,
    curveFrom: scalePoint(parsePoint(curvePoint.curveTo), frame),
    curveTo: scalePoint(parsePoint(curvePoint.curveFrom), frame),
    point: scalePoint(parsePoint(curvePoint.point), frame),
  };
}

export function encodeCurvePoint(
  curvePoint: DecodedCurvePoint,
  frame: Rect,
): Sketch.CurvePoint {
  return {
    ...curvePoint,
    hasCurveFrom: curvePoint.hasCurveTo,
    hasCurveTo: curvePoint.hasCurveFrom,
    curveFrom: stringifyPoint(unscalePoint(curvePoint.curveTo, frame)),
    curveTo: stringifyPoint(unscalePoint(curvePoint.curveFrom, frame)),
    point: stringifyPoint(unscalePoint(curvePoint.point, frame)),
  };
}

function curveTo(
  path: Path,
  current: DecodedCurvePoint,
  next: DecodedCurvePoint,
) {
  const curveTo = current.hasCurveTo ? current.curveTo : current.point;
  const curveFrom = next.hasCurveFrom ? next.curveFrom : next.point;
  const nextPoint = next.point;

  path.cubicTo(
    curveTo.x,
    curveTo.y,
    curveFrom.x,
    curveFrom.y,
    nextPoint.x,
    nextPoint.y,
  );
}

function lineTo(path: Path, next: DecodedCurvePoint) {
  const nextPoint = next.point;

  path.lineTo(nextPoint.x, nextPoint.y);
}

function curveOrLineTo(
  path: Path,
  current: DecodedCurvePoint,
  next: DecodedCurvePoint,
) {
  if (
    current.curveMode === Sketch.CurveMode.Straight &&
    next.curveMode === Sketch.CurveMode.Straight
  ) {
    lineTo(path, next);
  } else {
    curveTo(path, current, next);
  }
}

type CurveInfo = {
  radius: number;
  step: Point;
  drawSegment: boolean;
};

function stepToCurve(
  path: Path,
  current: DecodedCurvePoint,
  curveInfo: CurveInfo,
) {
  const currentPoint = current.point;

  const { radius, step } = curveInfo;

  if (radius > 0) {
    path.arcToTangent(
      currentPoint.x,
      currentPoint.y,
      currentPoint.x + step.x,
      currentPoint.y + step.y,
      radius,
    );
  } else {
    path.lineTo(currentPoint.x + step.x, currentPoint.y + step.y);
  }
}

export function path(
  CanvasKit: CanvasKit,
  points: Sketch.CurvePoint[],
  frame: Rect,
  isClosed: boolean,
): Path {
  const path = new CanvasKit.Path();

  if (points.length < 2) return path;

  const curvePoints = points.map((point) => decodeCurvePoint(point, frame));

  // Curves with only 2 points don't support corner radius, so we handle this
  // case specially
  if (curvePoints.length === 2) {
    const [current, next] = curvePoints;

    path.moveTo(current.point.x, current.point.y);

    curveOrLineTo(path, current, next);

    if (isClosed) {
      curveOrLineTo(path, next, current);

      path.close();
    }

    return path;
  }

  // Create an array of point tuples, [prev, current, next]
  const pairs = rotate(windowsOf(curvePoints, 3, true), -1);

  const curveInfoList: CurveInfo[] = pairs.map((pair) => {
    const [prev, current, next] = pair;

    const currentPoint = current.point;
    const nextPoint = next.point;
    const prevPoint = prev.point;

    const radius = getCornerRadius(
      prevPoint,
      currentPoint,
      nextPoint,
      current.cornerRadius,
    );

    const { step, drawSegment } = getStep(currentPoint, nextPoint, radius);

    return { radius, step, drawSegment };
  });

  zip(pairs, curveInfoList).forEach(([pair, curveInfo], index) => {
    const [, current, next] = pair;
    const isLast = index === pairs.length - 1;

    const currentPoint = current.point;
    const nextPoint = next.point;

    if (index === 0) {
      const { step } = curveInfo;

      path.moveTo(currentPoint.x + step.x, currentPoint.y + step.y);

      if (next.curveMode !== Sketch.CurveMode.Straight) {
        curveTo(path, current, next);
      }
    } else {
      if (
        // TODO: Handle one straight point with radius and one curve
        current.curveMode === Sketch.CurveMode.Straight &&
        next.curveMode === Sketch.CurveMode.Straight
      ) {
        stepToCurve(path, current, curveInfo);

        const { step, drawSegment } = curveInfo;

        if (drawSegment && (isClosed || !isLast)) {
          path.lineTo(nextPoint.x - step.x, nextPoint.y - step.y);
        }
      } else if (isClosed || !isLast) {
        curveTo(path, current, next);
      }
    }

    if (isClosed && isLast) {
      const [, current, next] = pairs[0];

      // If we have a corner radius, draw the final curve
      if (
        current.curveMode === Sketch.CurveMode.Straight &&
        next.curveMode === Sketch.CurveMode.Straight
      ) {
        stepToCurve(path, current, curveInfoList[0]);
      }

      // We need the path to be closed for line caps to draw correctly
      path.close();
    }
  });

  return path;
}

const getBorderPositionPathOp = (
  CanvasKit: CanvasKit,
  position: Sketch.BorderPosition,
): PathOp | undefined => {
  switch (position) {
    case Sketch.BorderPosition.Center:
      return;
    case Sketch.BorderPosition.Outside:
      return CanvasKit.PathOp.Difference;
    case Sketch.BorderPosition.Inside:
      return CanvasKit.PathOp.Intersect;
  }
};

export function getStrokedPath(
  CanvasKit: CanvasKit,
  path: Path,
  width: number,
  lineCap: Sketch.LineCapStyle,
  lineJoin: Sketch.LineJoinStyle,
  pathOp?: PathOp,
): Path {
  const copy = path.copy();

  // A 0-width path will never change the original when unioned
  if (width === 0 && pathOp === CanvasKit.PathOp.Union) return copy;

  if (
    !copy.stroke({
      width,
      cap: lineCapStyle(CanvasKit, lineCap),
      join: lineJoinStyle(CanvasKit, lineJoin),
    })
  ) {
    console.info('[getStrokedPath] Failed to stroke path');
    return copy;
  }

  if (pathOp !== undefined) {
    if (!copy.op(path, pathOp)) {
      console.info('[getStrokedPath] Failed to combine paths');
    }
  }

  return copy;
}

export function getStrokedBorderPath(
  CanvasKit: CanvasKit,
  path: Path,
  borderWidth: number,
  borderPosition: Sketch.BorderPosition,
  lineCap: Sketch.LineCapStyle,
  lineJoin: Sketch.LineJoinStyle,
) {
  switch (borderPosition) {
    case Sketch.BorderPosition.Center:
      break;
    case Sketch.BorderPosition.Outside:
    case Sketch.BorderPosition.Inside:
      borderWidth *= 2;
  }

  const pathOp = getBorderPositionPathOp(CanvasKit, borderPosition);

  return getStrokedPath(
    CanvasKit,
    path,
    borderWidth,
    lineCap,
    lineJoin,
    pathOp,
  );
}

export function unscaleCurvePoint(curvePoint: Sketch.CurvePoint, frame: Rect) {
  const unscale = (value: string) =>
    PointString.encode(unscalePoint(PointString.decode(value), frame));

  return {
    ...curvePoint,
    curveFrom: unscale(curvePoint.curveFrom),
    curveTo: unscale(curvePoint.curveTo),
    point: unscale(curvePoint.point),
  };
}

function pathCommandToSVGCommand(
  command: PathCommand,
): CommandWithoutQuadratics {
  switch (command[0]) {
    case PathCommandVerb.move: {
      const [, x, y] = command;
      return { type: 'move', to: { x, y } };
    }
    case PathCommandVerb.cubic: {
      const [, x1, y1, x2, y2, x3, y3] = command;
      return {
        type: 'cubicCurve',
        to: { x: x3, y: y3 },
        controlPoint1: { x: x1, y: y1 },
        controlPoint2: { x: x2, y: y2 },
      };
    }
    default:
      throw new Error('Path command not handled');
  }
}

export function pathToCurvePoints(
  path: Path,
  frame: Rect,
): Sketch.CurvePoint[] {
  const svgCommands = parsePathCmds(path.toCmds()).map(pathCommandToSVGCommand);

  // Assume a single path. `isClosed` should already be handled, before calling
  // this function, so we can ignore it here.
  const curvePoints = makePathsFromCommands(svgCommands)[0].points;

  return curvePoints.map((curvePoint) => unscaleCurvePoint(curvePoint, frame));
}

export function joinCurvePoints(
  segments: Sketch.CurvePoint[][],
  wrapsAround: boolean,
): Sketch.CurvePoint[] {
  const nonEmptySegments = segments.filter((segment) => segment.length > 0);

  if (nonEmptySegments.length === 0) return [];

  const [first, ...rest] = nonEmptySegments;

  const merged = rest.reduce((start, end) => {
    const lastStartPoint = start[start.length - 1];
    const firstEndPoint = end[0];

    return [
      ...start.slice(0, -1),
      SketchModel.curvePoint({
        curveMode: Sketch.CurveMode.Mirrored,
        hasCurveFrom: true,
        hasCurveTo: true,
        point: lastStartPoint.point,
        curveFrom: firstEndPoint.curveFrom,
        curveTo: lastStartPoint.curveTo,
      }),
      ...end.slice(1),
    ];
  }, first);

  if (wrapsAround) {
    const lastIndex = merged.length - 1;
    const last = merged[lastIndex];
    merged.splice(lastIndex, 1);
    merged[0].curveTo = last.curveTo;
  }

  return merged;
}

export function splitPath(path: Path, t: number): [Path, Path] {
  t = clamp(t, 0, 1);

  return [path.copy().trim(0, t, false)!, path.copy().trim(t, 1, false)!];
}
