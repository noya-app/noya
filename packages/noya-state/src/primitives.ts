import { Sketch } from '@noya-app/noya-file-format';
import { FontId, SYSTEM_FONT_ID } from '@noya-app/noya-fonts';
import {
  AffineTransform,
  Point,
  Rect,
  createBounds,
  distance,
  resize,
  transformRect,
} from '@noya-app/noya-geometry';
import type {
  CanvasKit,
  Paint,
  Path,
  RuntimeEffect,
  Shader,
  StrokeCap,
  StrokeJoin,
  TextAlign,
  TextStyle,
} from 'canvaskit';
import { CompassDirection } from 'noya-state';
import * as PathUtils from './primitives/path';
import {
  compassDirectionMap,
  getOppositeDirection,
  getRectExtentPoint,
} from './selection';

export * from './primitives/path';
export * from './primitives/pathCommand';

export function sign(number: number) {
  return number > 0 ? 1 : number < 0 ? -1 : 0;
}

export type ScalingOriginMode = 'extent' | 'center';

export type ScalingOptions = {
  constrainProportions: boolean;
  scalingOriginMode: ScalingOriginMode;
};

export function getScalingOptions({
  shiftKey,
  altKey,
}: {
  shiftKey: boolean;
  altKey: boolean;
}): ScalingOptions {
  return {
    constrainProportions: shiftKey,
    scalingOriginMode: altKey ? 'center' : 'extent',
  };
}

/**
 * Resize a rect in a compass direction
 */
export function resizeRect(
  rect: Rect,
  offset: Point,
  direction: CompassDirection,
  { scalingOriginMode, constrainProportions }: ScalingOptions,
): Rect {
  const oppositeDirection = getOppositeDirection(direction);

  const extent = getRectExtentPoint(rect, direction);
  const oppositeExtent = getRectExtentPoint(rect, oppositeDirection);

  if (scalingOriginMode === 'center') {
    offset = {
      x: offset.x * 2,
      y: offset.y * 2,
    };
  }

  const newExtent = { x: extent.x + offset.x, y: extent.y + offset.y };

  const multiplier = {
    x: compassDirectionMap[direction].x * 2 - 1,
    y: compassDirectionMap[direction].y * 2 - 1,
  };

  const newSize = {
    width: newExtent.x - oppositeExtent.x,
    height: newExtent.y - oppositeExtent.y,
  };

  const scaleX = (multiplier.x * newSize.width) / rect.width;
  const scaleY = (multiplier.y * newSize.height) / rect.height;

  const largestMagnitude =
    Math.abs(scaleX) > Math.abs(scaleY) ? scaleX : scaleY;

  const scale = constrainProportions
    ? { x: largestMagnitude, y: largestMagnitude }
    : {
        x: extent.x === oppositeExtent.x ? 1 : scaleX,
        y: extent.y === oppositeExtent.y ? 1 : scaleY,
      };

  // Adjust the sign of the scale:
  // - If scaling along a single axis, ensure a positive sign so we don't flip along the opposite axis
  // - If scaling opposite corners, ensure the rect remains in the correct quadrant
  if (extent.y === oppositeExtent.y) {
    scale.y = Math.abs(scale.y);
  } else if (sign(scale.y) !== sign(scaleY)) {
    scale.y *= -1;
  }

  if (extent.x === oppositeExtent.x) {
    scale.x = Math.abs(scale.x);
  } else if (sign(scale.x) !== sign(scaleX)) {
    scale.x *= -1;
  }

  switch (scalingOriginMode) {
    case 'extent':
      return transformRect(
        rect,
        AffineTransform.scale(scale.x, scale.y, oppositeExtent),
        false,
      );
    case 'center':
      const bounds = createBounds(rect);

      return transformRect(
        rect,
        AffineTransform.scale(scale.x, scale.y, {
          x: bounds.midX,
          y: bounds.midY,
        }),
        false,
      );
  }
}

export function point(point: Point): number[] {
  return [point.x, point.y];
}

export function color(CanvasKit: CanvasKit, color: Sketch.Color) {
  return CanvasKit.Color4f(color.red, color.green, color.blue, color.alpha);
}

export function rect<T extends Rect>(CanvasKit: CanvasKit, rect: T) {
  return CanvasKit.XYWHRect(rect.x, rect.y, rect.width, rect.height);
}

export function clearColor(CanvasKit: CanvasKit) {
  return CanvasKit.Color4f(0, 0, 0, 0);
}

export function shader(
  CanvasKit: CanvasKit,
  fill: Sketch.Fill | Sketch.Border,
  layerFrame: Rect,
  image?: ArrayBuffer,
  runtimeEffect?: RuntimeEffect,
  uniforms?: number[],
): Shader | undefined {
  switch (fill.fillType) {
    case Sketch.FillType.Shader:
      const fillScale =
        fill._class === 'fill'
          ? fill.patternFillType
          : Sketch.PatternFillType.Fill;

      const aspectRatio = layerFrame.width / layerFrame.height;
      const unitTransform = AffineTransform.multiply(
        AffineTransform.translate(layerFrame.x, layerFrame.y),
        fillScale === Sketch.PatternFillType.Stretch
          ? AffineTransform.identity
          : AffineTransform.scale(1 / aspectRatio, 1, { x: 0.5, y: 0.5 }),
      );

      const shader = runtimeEffect?.makeShader(
        uniforms ?? [],
        false,
        unitTransform.float32Array,
      );

      return shader;
    case Sketch.FillType.Color:
      const fillColor = fill.color
        ? color(CanvasKit, fill.color)
        : clearColor(CanvasKit);
      return CanvasKit.Shader.MakeColor(fillColor, CanvasKit.ColorSpace.SRGB);
    case Sketch.FillType.Gradient: {
      let colors: Float32Array[] = [];
      let positions: number[] = [];

      [...fill.gradient.stops]
        .sort((a, b) => a.position - b.position)
        .forEach((stop) => {
          colors.push(color(CanvasKit, stop.color));
          positions.push(stop.position);
        });

      const fromPoint = parsePoint(fill.gradient.from);
      const toPoint = parsePoint(fill.gradient.to);

      const aspectRatio = layerFrame.width / layerFrame.height;

      // CanvasKit draws gradients in absolute coordinates, while Sketch draws them
      // relative to the layer's frame. This function returns a matrix that converts
      // absolute coordinates into the range (0, 1).
      const unitTransform = AffineTransform.multiply(
        AffineTransform.translate(layerFrame.x, layerFrame.y),
        AffineTransform.scale(layerFrame.width, layerFrame.height),
      );

      switch (fill.gradient.gradientType) {
        case Sketch.GradientType.Linear: {
          const transformedFrom = unitTransform.applyTo(fromPoint);
          const transformedTo = unitTransform.applyTo(toPoint);

          return CanvasKit.Shader.MakeLinearGradient(
            point(transformedFrom),
            point(transformedTo),
            colors,
            positions,
            CanvasKit.TileMode.Clamp,
          );
        }
        case Sketch.GradientType.Radial: {
          const transformedCenter = unitTransform.applyTo(fromPoint);
          const transformedTo = unitTransform.applyTo(toPoint);

          const theta =
            Math.atan2(
              transformedTo.y - transformedCenter.y,
              transformedTo.x - transformedCenter.x,
            ) -
            Math.PI / 2;

          // We scale the coordinate system along the x axis to render an ellipse
          const coordinateSystemTransform = AffineTransform.rotate(
            theta,
            transformedCenter,
          ).scale(fill.gradient.elipseLength || 1, 1, transformedCenter);

          return CanvasKit.Shader.MakeRadialGradient(
            point(transformedCenter),
            distance(transformedTo, transformedCenter),
            colors,
            positions,
            CanvasKit.TileMode.Clamp,
            coordinateSystemTransform.float32Array,
          );
        }
        case Sketch.GradientType.Angular: {
          const hasStartPosition = positions[0] === 0;
          const hasEndPosition = positions[positions.length - 1] === 1;
          let rotationRadians = 0;

          // If the gradient has no start or end, we shift all the colors stops
          // to the beginning, and then rotate the gradient.
          //
          // We can't use the positions + angle parameters of MakeSweepGradient,
          // since these are clamped (and TileMode doesn't seem to affect this)
          if (!hasStartPosition && !hasEndPosition) {
            const startPosition = positions[0];
            positions = positions.map((p) => p - startPosition);
            colors.push(colors[0]);
            positions.push(1);

            rotationRadians = startPosition * 2 * Math.PI;
          } else if (hasEndPosition && !hasStartPosition) {
            colors.unshift(colors[colors.length - 1]);
            positions.unshift(0);
          } else if (hasStartPosition && !hasEndPosition) {
            colors.push(colors[0]);
            positions.push(1);
          }

          let matrix = AffineTransform.multiply(
            unitTransform,
            rotationRadians > 0
              ? AffineTransform.rotate(rotationRadians, { x: 0.5, y: 0.5 })
              : AffineTransform.identity,
            AffineTransform.scale(1 / aspectRatio, 1, { x: 0.5, y: 0.5 }),
          );

          return CanvasKit.Shader.MakeSweepGradient(
            0.5,
            0.5,
            colors,
            positions,
            CanvasKit.TileMode.Clamp,
            matrix.float32Array,
          );
        }
        default:
          return;
      }
    }
    case Sketch.FillType.Pattern: {
      if (!image || fill._class === 'border') return;

      const canvasImage = CanvasKit.MakeImageFromEncoded(image);

      if (!canvasImage) return;

      switch (fill.patternFillType) {
        case Sketch.PatternFillType.Tile: {
          return canvasImage.makeShaderCubic(
            CanvasKit.TileMode.Repeat,
            CanvasKit.TileMode.Repeat,
            1 / 3,
            1 / 3,
            CanvasKit.Matrix.multiply(
              CanvasKit.Matrix.translated(layerFrame.x, layerFrame.y),
              CanvasKit.Matrix.scaled(
                fill.patternTileScale,
                fill.patternTileScale,
              ),
            ),
          );
        }
        case Sketch.PatternFillType.Stretch:
        case Sketch.PatternFillType.Fit:
        case Sketch.PatternFillType.Fill: {
          const bounds = createBounds(layerFrame);
          const scaledRect = resize(
            {
              ...layerFrame,
              width: canvasImage.width(),
              height: canvasImage.height(),
            },
            layerFrame,
            fill.patternFillType === Sketch.PatternFillType.Stretch
              ? 'scaleToFill'
              : fill.patternFillType === Sketch.PatternFillType.Fit
              ? 'scaleAspectFit'
              : 'scaleAspectFill',
          );

          return canvasImage.makeShaderCubic(
            CanvasKit.TileMode.Decal,
            CanvasKit.TileMode.Decal,
            1 / 3,
            1 / 3,
            CanvasKit.Matrix.multiply(
              CanvasKit.Matrix.translated(
                bounds.midX - scaledRect.width / 2,
                bounds.midY - scaledRect.height / 2,
              ),
              CanvasKit.Matrix.scaled(
                scaledRect.width / canvasImage.width(),
                scaledRect.height / canvasImage.height(),
              ),
            ),
          );
        }
      }
    }
  }
}

export function fill(
  CanvasKit: CanvasKit,
  fill: Sketch.Fill | Sketch.Border,
  layerFrame: Rect,
  image?: ArrayBuffer,
  runtimeEffect?: RuntimeEffect,
  uniforms?: number[],
): Paint {
  const paint = new CanvasKit.Paint();

  switch (fill.fillType) {
    case Sketch.FillType.Color:
      paint.setColor(
        fill.color ? color(CanvasKit, fill.color) : clearColor(CanvasKit),
      );
      break;
    case Sketch.FillType.Gradient:
    case Sketch.FillType.Pattern:
    case Sketch.FillType.Shader: {
      const fillShader = shader(
        CanvasKit,
        fill,
        layerFrame,
        image,
        runtimeEffect,
        uniforms,
      );

      if (!fillShader) {
        paint.setColor(clearColor(CanvasKit));
        break;
      }

      paint.setShader(fillShader);
      paint.setAlphaf(fill.contextSettings.opacity);
      break;
    }
  }

  paint.setStyle(CanvasKit.PaintStyle.Fill);
  paint.setAntiAlias(true);

  return paint;
}

export function border(CanvasKit: CanvasKit, border: Sketch.Border): Paint {
  const paint = new CanvasKit.Paint();

  paint.setColor(
    border.color ? color(CanvasKit, border.color) : clearColor(CanvasKit),
  );
  paint.setStrokeWidth(border.thickness);
  paint.setStyle(CanvasKit.PaintStyle.Stroke);
  paint.setAntiAlias(true);

  return paint;
}

export function parsePoint(pointString: string): Point {
  const [x, y] = pointString.slice(1, -1).split(',');

  return {
    x: parseFloat(x),
    y: parseFloat(y),
  };
}

export function stringifyPoint({ x, y }: Point): string {
  return `{${x.toString()},${y.toString()}}`;
}

export function path(
  CanvasKit: CanvasKit,
  points: Sketch.CurvePoint[],
  frame: Rect,
  isClosed: boolean,
): Path {
  return PathUtils.path(CanvasKit, points, frame, isClosed);
}

export function textHorizontalAlignment(
  CanvasKit: CanvasKit,
  alignment: Sketch.TextHorizontalAlignment,
): TextAlign {
  switch (alignment) {
    case Sketch.TextHorizontalAlignment.Left:
      return CanvasKit.TextAlign.Left;
    case Sketch.TextHorizontalAlignment.Centered:
      return CanvasKit.TextAlign.Center;
    case Sketch.TextHorizontalAlignment.Right:
      return CanvasKit.TextAlign.Right;
    case Sketch.TextHorizontalAlignment.Justified:
      return CanvasKit.TextAlign.Justify;
    case Sketch.TextHorizontalAlignment.Natural: // What is this?
      return CanvasKit.TextAlign.Start;
  }
}

export function lineJoinStyle(
  CanvasKit: CanvasKit,
  lineJoin: Sketch.LineJoinStyle,
): StrokeJoin {
  switch (lineJoin) {
    case Sketch.LineJoinStyle.Bevel:
      return CanvasKit.StrokeJoin.Bevel;
    case Sketch.LineJoinStyle.Round:
      return CanvasKit.StrokeJoin.Round;
    case Sketch.LineJoinStyle.Miter:
      return CanvasKit.StrokeJoin.Miter;
  }
}

export function lineCapStyle(
  CanvasKit: CanvasKit,
  lineCap: Sketch.LineCapStyle,
): StrokeCap {
  switch (lineCap) {
    case Sketch.LineCapStyle.Butt:
      return CanvasKit.StrokeCap.Butt;
    case Sketch.LineCapStyle.Projecting:
      return CanvasKit.StrokeCap.Square;
    case Sketch.LineCapStyle.Round:
      return CanvasKit.StrokeCap.Round;
  }
}

export type SimpleTextDecoration = 'none' | 'underline' | 'strikethrough';

export function createCanvasKitTextStyle(
  CanvasKit: CanvasKit,
  fontId: FontId,
  attributes: Sketch.StringAttribute['attributes'],
  decoration: SimpleTextDecoration,
): TextStyle {
  const textColor = attributes.MSAttributedStringColorAttribute;
  const font = attributes.MSAttributedStringFontAttribute;

  return new CanvasKit.TextStyle({
    ...(textColor && { color: color(CanvasKit, textColor) }),
    fontFamilies: fontId
      ? [fontId.toString(), SYSTEM_FONT_ID]
      : [SYSTEM_FONT_ID],
    fontSize: font.attributes.size,
    letterSpacing: attributes.kerning,
    ...(decoration === 'none'
      ? {}
      : {
          decoration:
            decoration === 'underline'
              ? CanvasKit.UnderlineDecoration
              : CanvasKit.LineThroughDecoration,
          // There's currently a typo in the TypeScript types, "decration"
          ['decorationStyle' as any]: CanvasKit.DecorationStyle.Solid,
        }),
  });
}

export function pathOp(
  CanvasKit: CanvasKit,
  booleanOperation: Sketch.BooleanOperation,
) {
  switch (booleanOperation) {
    case Sketch.BooleanOperation.None:
      return undefined;
    case Sketch.BooleanOperation.Union:
      return CanvasKit.PathOp.Union;
    case Sketch.BooleanOperation.Subtract:
      return CanvasKit.PathOp.Difference;
    case Sketch.BooleanOperation.Intersection:
      return CanvasKit.PathOp.Intersect;
    case Sketch.BooleanOperation.Difference:
      return undefined;
  }
}

export function pathFillType(
  CanvasKit: CanvasKit,
  windingRule: Sketch.WindingRule,
) {
  switch (windingRule) {
    case Sketch.WindingRule.EvenOdd:
      return CanvasKit.FillType.EvenOdd;
    case Sketch.WindingRule.NonZero:
      return CanvasKit.FillType.Winding;
  }
}
