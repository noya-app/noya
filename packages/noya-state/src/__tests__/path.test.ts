import type { CanvasKit as CanvasKitType, Path } from 'canvaskit';
import { loadCanvasKit } from 'noya-renderer';
import { SketchModel } from 'noya-sketch-model';
import { PathCommand, PathCommandVerb, parsePathCmds } from 'noya-state';
import { round } from 'noya-utils';
import { path } from '../primitives/path';

let CanvasKit: CanvasKitType;

beforeAll(async () => {
  CanvasKit = await loadCanvasKit();
});

function describePathCommand([verb, ...args]: PathCommand) {
  return `${PathCommandVerb[verb]}(${args.map((n) => round(n, 2)).join(', ')})`;
}

function describePath(path: Path) {
  return parsePathCmds(path.toCmds()).map(describePathCommand);
}

test('rectangle', () => {
  const rectangle = SketchModel.rectangle({
    frame: SketchModel.rect({ x: 0, y: 0, width: 100, height: 100 }),
  });

  const result = path(
    CanvasKit,
    rectangle.points,
    rectangle.frame,
    rectangle.isClosed,
  );

  expect(describePath(result)).toMatchSnapshot();
});

test('rounded rectangle', () => {
  const rectangle = SketchModel.rectangle({
    frame: SketchModel.rect({ x: 0, y: 0, width: 100, height: 100 }),
    fixedRadius: 10,
  });

  const result = path(
    CanvasKit,
    rectangle.points,
    rectangle.frame,
    rectangle.isClosed,
  );

  expect(describePath(result)).toMatchSnapshot();
});

test('oval', () => {
  const oval = SketchModel.oval({
    frame: SketchModel.rect({ x: 0, y: 0, width: 100, height: 100 }),
  });

  const result = path(CanvasKit, oval.points, oval.frame, oval.isClosed);

  expect(describePath(result)).toMatchSnapshot();
});
