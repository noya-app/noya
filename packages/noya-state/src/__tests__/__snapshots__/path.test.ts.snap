// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`oval 1`] = `
Array [
  "move(50, 100)",
  "cubic(77.61, 100, 100, 77.61, 100, 50)",
  "cubic(100, 22.39, 77.61, 0, 50, 0)",
  "cubic(22.39, 0, 0, 22.39, 0, 50)",
  "cubic(0, 77.61, 22.39, 100, 50, 100)",
  "close()",
]
`;

exports[`rectangle 1`] = `
Array [
  "move(0, 0)",
  "line(100, 0)",
  "line(100, 100)",
  "line(100, 100)",
  "line(0, 100)",
  "line(0, 100)",
  "line(0, 0)",
  "line(0, 0)",
  "close()",
]
`;

exports[`rounded rectangle 1`] = `
Array [
  "move(10, 0)",
  "line(90, 0)",
  "conic(100, 0, 100, 10, 0.71)",
  "line(100, 90)",
  "line(100, 90)",
  "conic(100, 100, 90, 100, 0.71)",
  "line(10, 100)",
  "line(10, 100)",
  "conic(0, 100, 0, 90, 0.71)",
  "line(0, 10)",
  "line(0, 10)",
  "conic(0, 0, 10, 0, 0.71)",
  "close()",
]
`;
